File: bls-server/src/middleware/http-metrics.ts
/**
 * HTTP 请求指标中间件
 *
 * 自动采集所有 API 请求的：
 * - 请求数（counter）
 * - 请求耗时（histogram）
 * - 错误数（counter）
 */

import type { Context, Next } from 'koa';
import { httpRequestsTotal, httpRequestDurationSeconds, httpRequestErrorsTotal } from '../observability/metrics';

export async function httpMetricsMiddleware(ctx: Context, next: Next): Promise<void> {
  const start = Date.now();

  try {
    await next();
  } finally {
    const duration = (Date.now() - start) / 1000;
    const method = ctx.method;
    // 标准化路由（避免高基数）：用 ctx._matchedRoute 或 fallback 到 path
    const route = (ctx._matchedRoute ?? ctx.path) || '/unknown';
    const status = String(ctx.status);

    httpRequestsTotal.inc({ method, route, status });
    httpRequestDurationSeconds.observe({ method, route }, duration);
    if (ctx.status >= 400) {
      httpRequestErrorsTotal.inc({ method, route });
    }
  }
}

File: bls-rust-server/src/middleware/http_metrics.rs
use std::time::Instant;

use axum::extract::{Request, State};
use axum::middleware::Next;
use axum::response::Response;

use crate::state::AppState;

pub async fn http_metrics(
    State(_state): State<AppState>,
    request: Request<axum::body::Body>,
    next: Next,
) -> Response {
    let method = request.method().to_string();
    let path = normalize_path(request.uri().path());
    let start = Instant::now();
    let response = next.run(request).await;
    let status = response.status().as_u16().to_string();
    crate::observability::metrics::http_requests_total()
        .with_label_values(&[&method, &path, &status])
        .inc();
    crate::observability::metrics::http_request_duration_seconds()
        .with_label_values(&[&method, &path])
        .observe(start.elapsed().as_secs_f64());
    response
}

fn normalize_path(path: &str) -> String {
    if path.starts_with("/api/system/") || path.starts_with("/api/v1/system/") {
        let parts: Vec<&str> = path.split('/').collect();
        if parts.len() > 4 {
            return parts[..4].join("/");
        }
    }
    path.to_string()
}
